src/primitives/iter.rs

```diff
@@ -0,0 +1,414 @@
+                }
+                Greater => {
+                    self.last_byte = self.iter.next();
+                    let next = self.last_byte.unwrap_or(0);
+                    Some(Fe32(((last << (bit_offset - 3)) | (next >> (11 - bit_offset))) & 0x1f))
+                }
+            }
+        } else {
+            None
+        }
+    }
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        let (min, max) = self.iter.size_hint();
+        let (min, max) = match self.last_byte {
+            // +1 because we set last_byte with call to `next`.
+            Some(_) => (min + 1, max.map(|max| max + 1)),
+            None => (min, max),
+        };
+
+        let min = bytes_len_to_fes_len(min);
+        let max = max.map(bytes_len_to_fes_len);
+
+        (min, max)
+    }
+}
+
+/// The number of fes encoded by n bytes, rounded up because we pad the fes.
+fn bytes_len_to_fes_len(bytes: usize) -> usize {
+    let bits = bytes * 8;
+    (bits + 4) / 5
+}
+
+/// Extension trait for field element iterators
+pub trait Fe32IterExt: Sized + Iterator<Item = Fe32> {
+    /// Adapts the Fe32 iterator to output bytes instead.
+    ///
+    /// If the total number of bits is not a multiple of 8, any trailing bits
+    /// are simply dropped.
+    fn fes_to_bytes(mut self) -> FeToByteIter<Self> {
+        FeToByteIter { last_fe: self.next(), bit_offset: 0, iter: self }
+    }
+
+    /// Adapts the Fe32 iterator by prepending `Fe::32:Q` (witness version 0).
+    fn with_witness_v0(self) -> WitnessVersionIter<Self> { self.with_witness_version(Fe32::Q) }
+
+    /// Adapts the Fe32 iterator by prepending `Fe::32:P` (witness version 1).
+    fn with_witness_v1(self) -> WitnessVersionIter<Self> { self.with_witness_version(Fe32::P) }
+
+    /// Adapts the Fe32 iterator by prepending `fe` (witness version).
+    ///
+    /// Accepts any `Fe32`, does no checks on the validity of `witness_version`.
+    fn with_witness_version(self, witness_version: Fe32) -> WitnessVersionIter<Self> {
+        WitnessVersionIter { witness_version: Some(witness_version), iter: self }
+    }
+
+    /// Adapts the Fe32 iterator to append a checksum to the end of the data.
+    ///
+    /// Because the HRP of a bech32 string needs to be expanded before being
+    /// checksummed, this iterator is a little bit inconvenient to use on raw
+    /// data. The [`ChecksumIter::with_checksummed_hrp`] methods may be of use.
+    fn checksum<Ck: Checksum>(self) -> ChecksumIter<Self, Ck> {
+        ChecksumIter {
+            iter: self,
+            checksum_remaining: Ck::CHECKSUM_LENGTH,
+            checksum_engine: checksum::Engine::new(),
+        }
+    }
+
+    /// Adapts the Fe32 iterator to output characters using `hrp` for the human-readable part.
+    ///
+    /// Note, `hrp` is expected to be the same as that fed into the checksum engine with
+    /// `with_checksummed_hrp`.
+    fn hrp_char(self, hrp: &Hrp) -> HrpCharIter<'_, Self> {
```

Comment:
Is there any reason to keep `with_checksummed_hrp` and `hrp_char` as separate functions? Why not combine into a single adapter? Are there reasonable use cases where these two functions are not always called together?

Reply:
As an iterator adaptor, `hrp_char` doesn't know that there's a checksum engine underneath that it could be feeding data into. So it just outputs the data but doesn't do anything about the checksum.

Meanwhile `with_checksummed_hrp` is a method specifically on `ChecksumIter`. So it can feed data into the checksum engine, but it doesn't output it, presumably because `hrp_char` will do it for you.

Reply:
Ah, the other distinction is that `HrpCharIter` yields characters while `ChecksumIter` yields `Fe32`s. This is probably why we have separate functions.

Reply:
My vote is:

* Drop `hrp_char` from `Fe32IterExt`. This is a confusing method to have on a general Fe32 iterator
* Change `with_checksummed_hrp` to yield a new iterator type `HrpChecksumIter` which yields characters rather than Fe32s, yields the HRP as character data, and simultaneously feeds it into the underlying checksum engine
* Rename `with_checksummed_hrp` to just `with_hrp`

Good call complaining about this :)

Reply:
This is an interesting question, the fact that the last 3 calls in this combinator chain are all segwit specific has been bothering me. Also I don't see any use case where one leaves out any of them, basically one either iterates over the fes returned by `bytes_to_fes` (can map to chars) or else one adds the hrp, checksum etc.
```rust
//! let iterator = witness_prog
//!     .iter()
//!     .copied() // Iterate over bytes.
//!     .bytes_to_fes() // Convert bytes to field elements in-line.
//!     .with_witness_v0() 
//!     .checksum::<Bech32>() // Convert to a [`ChecksumIter`] (append a bech32 checksum).
//!     .with_checksummed_hrp(&hrp) // Feed HRP into the checksum.
//!     .hrp_char(&hrp); // Turn the fe stream into a char stream with HRP.
//! let hrpstring: String = iterator.collect();
```

We could have a different design where one adapts an `Fe32` iterator straight into an iterator that iterates over chars, and includes everything (hrp, checksum etc.). All the bip test vectors and `lib.rs` use cases are possible if one chains the witness version onto the front of the fe32 iterator
```rust
let iterator = iter::once(Fe32::Q)
    .chain(
        witness_prog
            .iter()
            .copied()
            .bytes_to_fes()
    )
    .char_iter::<Bech32>(hrp)
```

Note that once we add the final `segwit` API no-one should ever have to write the `iter::once` chain themselves.

Reply:
hmmm, that `chain` is ugly, it may be better to have the witness version as a parameter on `char_iter`.

Reply:
`.checksum::<Bech32>()` is not segwit specific. It takes a stream of field elements and suffixes a given checksum. It's one of the most general things that this library does.

I think it's reasonable to combine `.with_checksummed_hrp` and `.hrp_char` because these methods really don't make sense without each other (sometimes you do want an already-constructed checksum engine, but this is fairly rare, and for this we should just provide a separate `.checksum_with_engine()` adaptor).

But `.checksum()` is really important whenever you're working with field elements and don't want to convert them to characters. i.e. literally any use of field elements except for display. Including verifying checksums.

> We could have a different design where one adapts an `Fe32` iterator straight into an iterator that iterates over chars, and includes everything (hrp, checksum etc.). All the bip test vectors and `lib.rs` use cases are possible if one chains the witness version onto the front of the fe32 iterator

I really strongly dislike this and I'd fork the library if that's what we would up doing.

> All the bip test vectors and `lib.rs` use cases are possible if one chains the witness version onto the front of the fe32 iterator

Yes, but `.chain` absolutely sucks, as you can see, which is why we have `WitnessVersionIterator`.

Reply:
Oh shiiiit, this idea turned out super nice - pushing as a separate patch you guys to look at. Will squash if it gets a concept ack.

Reply:
If I have a vector of `Fe32`s with a checksum at the end, how can I verify that checksum using this API?

Reply:
One could write something like this (do you want me to add this, I didn't test it for correctness):
```rust

fn validate_checksum<Ck: Checksum>(v: Vec<Fe32>) -> bool {
    if v.len() < Ck::CHECKSUM_LENGTH {
        panic!("to short");
    }
    let chk_boundry = v.len() - Ck::CHECKSUM_LENGTH;
    let data = &v[..chk_boundry];
    let chk = &v[chk_boundry..];

    let mut engine = checksum::Engine::<Ck>::new();
    for fe in data.iter() {
        engine.input_fe(*fe);
    }
    engine.input_target_residue();

    let mut checksum_remaining = Ck::CHECKSUM_LENGTH;
    for c in chk.iter() {
        if *c != Fe32(engine.residue().unpack(checksum_remaining)) {
            return false
        }
        checksum_remaining -= 1;
    }
    true
}
```

Reply:
Although note that currently validating check sums is no where in the code because its in `hrpstring`.

Reply:
So I wrote up some code how to do it then finally realized you mean that by removing the `ChecksumIter` I'm making this hard. Perhaps we should keep the `iter` module as you originally had it and add the new `CharIter` on top of it so that users have both options? But that does lead us back to @clarkmoody's question, "is there a usecase for only `with_checksummed_hrp` and not `hrp_char`"? From an engineering perspective, its not trivial to combine these two adaptors (at least I don't see an easy way).

Reply:
`with_checksummed_hrp` is useful for verifying checksums. Then you don't want to convert to char, you want to run all your data through it and check the residue of the underlying engine.

Let's switch to Signal for a bit and see if we can hash something out.

Reply:
For the record -- @tcharding and I had a phone call and agreed that his approach is actually correct here. Separately, we discussed #112 and the `data_iter` method there, and realized that we needed to merge the two PRs because all these "what are the proper iterator APIs" questions need to be answered at once if Clark is to have a fair chance at reviewing it.